### ISSUE TITLE
Which spreadsheet values should map to incident:adminlevels?

### ISSUE THREAD

**Issue by jeancochrane** [security-force-monitor/sfm-cms#650]
I'm closing in on the importer bug that has been blocking the recreation of our import. Fixing it will require adjusting the way that we import `incident:adminlevel1` and `incident:adminlevel2` but I'm not sure I fully understand what expected behavior should be.

Right now, we retrieve the OSM ID stored in column Z (`City or smallest administrative unit GeonameID`) and use it to construct the admin levels for both `adminlevel1` and `adminlevel2`. We do this by inferring the hierarchy from our raw OSM data:

https://github.com/security-force-monitor/sfm-cms/blob/98c430831ce9c5f842d32118150915d790018474/sfm_pc/management/commands/import_google_doc.py#L2045-L2052

However, the spreadsheet also has a column `ADMIN1 GeonameID` that isn't used at all. This makes me think that perhaps we're inferring at least one value here when we really should be reading it from the sheet. Which portions of the spreadsheet should ultimately map to `incident:adminlevel1` and `incident:adminlevel2` @tlongers?

The underlying bug is a failure to properly save `admin1` and `admin2` on the model, so once we settle where those values should come from the fix should be pretty quick.

**Comment by jeancochrane** [comment]
Blocking #651.

**Comment by tlongers** [comment]
Good catch. 

The aim should be to completely remove ambiguity as regards where we get `admin1` and `admin2` values from. 

In my mind, this means simplifying the issue by removing any automatic assignment and enabling system users to specify what locational information should be assigned to which tier:

 - When importing data, then, this means taking the OSM ID for `admin1` from the column `ADMIN1 GeonameID` rather than inferring it. We also currently specify one or both of these values wherever a location is linked to a unit too.
 - Users adding a location to an incident through the app itself can currently connect an OSM object directly to each of the three levels of precision (`exact location`, `settlement`, `top administrative areas`). 

This latter pattern, of manual assignment of OSM objects to different levels, has more promise I think. However, it is not implemented with respect to how we link locations to units. My feel is that it should be though. Where are you with this?

**Comment by jeancochrane** [comment]
> When importing data, then, this means taking the OSM ID for admin1 from the column ADMIN1 GeonameID rather than inferring it. We also currently specify one or both of these values wherever a location is linked to a unit too.

Got it! And which column of the spreadsheet should we use for `admin2`? Is that what `City or smallest administrative unit GeonameID` represents?

> This latter pattern, of manual assignment of OSM objects to different levels, has more promise I think. However, it is not implemented with respect to how we link locations to units. My feel is that it should be though.

Absolutely agree. I think we should treat this as a separate issue, though, since it implicates data modeling changes and data entry UI changes in addition to importer changes. Do we have an issue open for it yet?

**Comment by tlongers** [comment]
> And which column of the spreadsheet should we use for admin2? Is that what `City or smallest administrative unit GeonameID` represents?

Yes. I'd like @tonysecurityforcemonitor to confirm quickly.

>Do we have an issue open for it yet?
No, I'll open one up.

**Comment by tonysecurityforcemonitor** [comment]
Sorry - I'm a little lost, and I may be confusing things more with the response below:

It sounds like the question is about the what we should do to replace the inference system currently in place which infers: 
`admin1` (which generally is `admin_level=4` in OSM)
`admin2` (which generally is `admin_level=6` in OSM)

`City or smallest administrative unit GeonameID` in most cases is an OSM node. @tlongers is right to point out that `admin1` can be manually pulled from the sheet from column `ADMIN1 GeonameID` these will always match --- but there is a no equivalent column for `admin2` in the sheet, and no equivalent in the data model more generally. In most cases where we have a town/settlement listed as the location of an incident we would put that in `City or smallest administrative unit GeonameID` which would be a node. In some cases the source would state a more abstract location which would be `admin_level=6 in OSM - such as this case:https://whowasincommand.com/en/violation/view/b53e4059-7b74-481a-8638-7352d148f3aa/

Basically, `City or smallest administrative unit GeonameID` is doing heavy lifting of representing a geographic entity that is smaller than `admin1` or `ADMIN1 GeonameID` or `admin_level=4` in OSM, which could be a node, relation or way, and may, or may not, line up with `admin2`

HOWEVER - I'm not sure I'm exactly answering the question that was asked here.

**Comment by tlongers** [comment]
Thanks @tonysecurityforcemonitor, that's helpful.

@jeancochrane Perhaps the better question is "What is the platform/importer currently doing with the data from `City or smallest administrative unit GeonameID`", and is this different from what is being done with `admin2`"?

**Comment by jeancochrane** [comment]
Currently, the importer assigns `City or smallest administrative unit GeonameID` to the attribute `Voilation.osmid`, which as far as I can tell is not displayed anywhere in the app or exposed in any data entry views. It also retrieves the OSM geometry for this ID, and uses it to infer `Violation.adminlevel1` and `Violation.adminlevel2`.

For context, here are the labels we use for different geometry attributes in the Incident data entry form:

- "Exact Location": `violation.location`
- "Settlement": `Violation.adminlevel1`
- "Top  administrative area": `Violation.adminlevel2`

Does that help clarify things? I still feel uncertain what the proper way of populating `adminlevel2` is. I'm open to the idea that maybe the data model in the app is misaligned with the data model expected by the importer, as suggested by @tonysecurityforcemonitor's comment.

**Comment by tlongers** [comment]
I think I've got to the root of the app's behaviour here. Here's how the different parts of the location information we provide in the spreadsheet map on to the app:

Import sheet|WWIC read view|WWIC data entry view|OSM object type|Geometry type
---|---|---|---|---
Location|"Location" field in header; descriptive sentence beneath map|"Location Description"|N/A|N/A|
Site: Exact Location - Longitude or OSM Node Name|Not present|Not present|Node/lon|point
Site: Exact Location - Latitude or OSM Node ID|Not present|Not present|Node/lat|point
City or smallest administrative unit GeoName|Pin in map area|"Exact Location" field|node/way/relation|point/polygon|
City or smallest administrative unit GeoNameID|Pin in map area|"Exact Location" field|node/way/relation|point/polygon|
ADMIN1 Geoname|Descriptive sentence beneath map|"Top administrative area"|Relation|polygon
ADMIN1 GeonameID|Descriptive sentence beneath map|"Top administrative area"|Relation|polygon
Event division id|"Country" field in header; Descriptive sentence beneath map|Not present in location edit/create, but is present in Incident basics.|N/A|N/A/
Not present in spreadsheet|Not used in read view|"Settlement"|Relation|Polygon|

More verbosely:

The data in the spreadsheet field pair `Site: Exact Location` is not being used at all in the app; it doesn't seem to be imported.

The data in the spreadsheet field pair `City or smallest administrative unit` are being used to populate "Exact Location" in the backend. This is point geometry, and used to create the map pin in the read-view. 

The app's field `admin1` does not have an equivalent in the spreadsheet: it is derived from the OSM name and node id contained in the `City or smallest administrative unit` field. If the app (possibly through the location importer) is not able to infer an `admin1`, then this field will be blank. For example in incident `7fd503ec-a76d-4921-a522-2cffa67dd6d8` the value pair for `City or smallest administrative unit` is the node `Maiduguri, 27565111`. The app infers the correct OSM Admin Level 6 boundary for the node as `Maiduguri, 3721188` and places it in the `Settlement` field. However, this doesn't work in all cases: in `1e52b1ac-cb62-455f-87b6-01c66c381fe3` the value pair for `City or smallest administrative unit` is the node `Isaka, 501496639` (which is placed in "Exact Location" in the backend). The relevant OSM Admin Level 6 boundary, according to Nominatim, is the relation `Okrika, 3720761` - but the app does not find this. 

So, to answer the question Jean originally asked before we got sucked into the vortex:

-  `admin2` values should come from the spreadsheet field pair `ADMIN1 Geoname`.
- `admin1` has no direct equivalent in the spreadsheet. If can't be inferred by the app, allow it to be blank.

Additional observations:
- In this context, we have give different location terminology sets in use: OSM's object types, OSM's administrative levels, the way these admin levels are represented in the app, the spreadsheet headings, the front end labelling. No wonder it's difficult!
- the spreadsheet entries for the  "Exact Location" field pair are intentionally excluded.
- The ultimate answer, as I pitched earlier, is to refactor the locations around a more coherent model that we can exercise better control over exactly what geometry goes where.

**Comment by jeancochrane** [comment]
Thanks for digging in here @tlongers, I really appreciate this level of detail.

Some thoughts:

> The ultimate answer, as I pitched earlier, is to refactor the locations around a more coherent model that we can exercise better control over exactly what geometry goes where.

Strong agree!

> the spreadsheet entries for the "Exact Location" field pair are intentionally excluded.

What indicates this to you? Looking at the importer code, it seems on first glance like we're using the Exact Location field in the spreadsheet as I would expect:

https://github.com/security-force-monitor/sfm-cms/blob/98c430831ce9c5f842d32118150915d790018474/sfm_pc/management/commands/import_google_doc.py#L2000-L2021

> So, to answer the question Jean originally asked before we got sucked into the vortex:
>
> - `adminlevel2` values should come from the spreadsheet field pair `ADMIN1 Geoname`.
> - `adminlevel1` has no direct equivalent in the spreadsheet. If can't be inferred by the app, allow it to be blank.

Can do. Do we need to update data entry views as part of this work, too? It seems to me at the very least that `Settlement` is the wrong word for `adminlevel2` and we should probably label it differently.

**Comment by tlongers** [comment]
>What indicates this to you? Looking at the importer code, it seems on first glance like we're using the Exact Location field in the spreadsheet as I would expect ....

When reviewing where values from the spreadsheet appeared in WWIC, I couldn't find anywhere that the values from the `Exact Location` pair were used.  

For example, this record on staging`~en/violation/edit/4c3742cd-3ed6-48cc-b682-8825c33d9e9c/`. The `Exact Location` field pair in the spreadsheet contains the values `Giwa Barracks, 196204935`, but they are not present in the read or edit view of the record. 

Could it be that they are processed, but not brought into the user-facing bits of the app anywhere?

> Can do. Do we need to update data entry views as part of this work, too? It seems to me at the very least that Settlement is the wrong word for adminlevel2 and we should probably label it differently

Yes, you're right. In the data entry side, these labels seem appropriate:

`Exact Location` --> `Settlement`
`Settlement` --> `Settlement administrative area`

**Comment by jeancochrane** [comment]
I made the above changes to Violation admin levels in https://github.com/security-force-monitor/sfm-cms/commit/51bbcde892c5ad3666e59e78f5bbd101bb6f1f74 and https://github.com/security-force-monitor/sfm-cms/commit/2c354a5f4d917ae35627c7c38f981be5d3d5a2d5. I'll kick off the import on staging again this afternoon and report back tomorrow if all looks good!

Digging into `Exact Locations` some more, I believe @tlongers is right that we're not saving them properly in the importer. The code [I referenced above](https://github.com/security-force-monitor/sfm-cms/issues/650#issuecomment-537528235) was saving values to `Violation.LocationName` and `Violation.LocationId`, which are actually deprecated `Exact Location` attributes that are no longer used -- the relevant code for importing `violation:location` below demonstrates that we indeed use the value of `City or smallest administrative area`, aliased to `geo`:

https://github.com/security-force-monitor/sfm-cms/blob/51bbcde892c5ad3666e59e78f5bbd101bb6f1f74/sfm_pc/management/commands/import_google_doc.py#L2103-L2109

Two options for addressing this are:

1. Open a new issue for this discrepancy and fix it
2. Wait for a more comprehensive refactor of the Violation Location models

Since the discrepancy isn't blocking the import or search index, the way that this issue was, I'm fine with either path -- it's your call @tlongers.
